### ISSUE TITLE
Webpack 5 uses dramatically more memory than Webpack 4

### ISSUE THREAD

**Issue by andersk** [webpack/webpack#13550]
# Bug report

**What is the current behavior?**

We’ve been unable to migrate Zulip from webpack 4 to webpack 5 due to an enormous memory usage regression. Memory usage is important to us because many of our users are on VMs with just 2 GB RAM.

In an effort to debug this with as few variables and moving parts as possible, I deleted everything except our JavaScript and an extremely minimal nucleus of the webpack configuration. Then I ran each version of webpack several times, deleting the output and cache directories before each run. I observed that webpack 5 uses nearly 4 GB of memory on this benchmark, while webpack 4 used just 500 to 800 MB.

```js
// webpack.config.js
const assets = require("./tools/webpack.assets.json");

module.exports = {
    mode: "production",
    context: __dirname,
    entry: assets,
};
```

**If the current behavior is a bug, please provide the steps to reproduce.**

Clone [this branch](https://github.com/andersk/zulip/tree/webpack-benchmark), and observe the `maxresident` values from `\time webpack`:

```console
$ git clone -b webpack-benchmark --depth=1 https://github.com/andersk/zulip.git
$ cd zulip
$ yarn
$ node_modules/.bin/webpack --version
webpack 5.38.1
webpack-cli 4.7.2
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
311.76user 3.99system 0:36.50elapsed 864%CPU (0avgtext+0avgdata 3985384maxresident)k
7792inputs+23736outputs (0major+1287196minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
310.15user 3.76system 0:36.20elapsed 867%CPU (0avgtext+0avgdata 3805580maxresident)k
0inputs+23672outputs (0major+1263156minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
311.75user 3.95system 0:36.73elapsed 859%CPU (0avgtext+0avgdata 3918600maxresident)k
0inputs+23672outputs (0major+1284466minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
313.17user 3.91system 0:36.71elapsed 863%CPU (0avgtext+0avgdata 3819140maxresident)k
0inputs+23704outputs (0major+1263395minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
310.46user 3.89system 0:36.15elapsed 869%CPU (0avgtext+0avgdata 3887100maxresident)k
0inputs+23704outputs (0major+1283419minor)pagefaults 0swaps
$ yarn add webpack@^4
$ node_modules/.bin/webpack --version
webpack 4.46.0
webpack-cli 4.7.2
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
234.98user 4.59system 0:33.39elapsed 717%CPU (0avgtext+0avgdata 658272maxresident)k
2072inputs+52072outputs (0major+1500173minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
235.43user 4.37system 0:33.13elapsed 723%CPU (0avgtext+0avgdata 519372maxresident)k
0inputs+52016outputs (0major+1438493minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
235.23user 4.36system 0:32.77elapsed 731%CPU (0avgtext+0avgdata 784472maxresident)k
0inputs+52008outputs (0major+1423157minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
237.57user 4.58system 0:33.16elapsed 730%CPU (0avgtext+0avgdata 780992maxresident)k
0inputs+52080outputs (0major+1487320minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
235.55user 4.48system 0:33.06elapsed 725%CPU (0avgtext+0avgdata 775600maxresident)k
0inputs+51984outputs (0major+1493409minor)pagefaults 0swaps
```

**What is the expected behavior?**

webpack 5 should have similar memory usage to webpack 4.

**Other relevant information:**
webpack version:  5.38.1
Node.js version: 14.17.0
Operating System: NixOS 21.11
Additional tools: yarn 1.22.10

**Comment by myNameIsDu** [comment]
Webpack5 cache is in memory by default in development mode, [you can disable the cache with false](https://webpack.js.org/configuration/other-options/#cache),[ or put it on disk](https://webpack.js.org/configuration/other-options/#cachetype)

**Comment by andersk** [comment]
@myNameIsDu As you can see, I’m in production mode, so the cache is already disabled. (I tried `cache: false` anyway, and as expected, it has no effect.)

**Comment by andersk** [comment]
By bisecting commits in webpack and terser-webpack-plugin, I was able to attribute most of the memory usage regression to webpack-contrib/terser-webpack-plugin#344.

I found that patching terser-webpack-plugin to remove `enableWorkerThreads: true` reduces memory usage back to the range of Webpack 4, without a noticeable performance impact.

@alexander-akait FYI.

```console
$ node_modules/.bin/webpack --version
webpack 5.38.1
webpack-cli 4.7.2
$ sed -i '/enableWorkerThreads/d' node_modules/terser-webpack-plugin/dist/index.js
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack-cli --stats=errors-only
311.26user 5.43system 0:36.25elapsed 873%CPU (0avgtext+0avgdata 675108maxresident)k
0inputs+23704outputs (0major+1511627minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack-cli --stats=errors-only
312.24user 5.60system 0:36.35elapsed 874%CPU (0avgtext+0avgdata 671892maxresident)k
0inputs+23704outputs (0major+1585726minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack-cli --stats=errors-only
312.78user 5.36system 0:36.48elapsed 871%CPU (0avgtext+0avgdata 675744maxresident)k
0inputs+23704outputs (0major+1549216minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack-cli --stats=errors-only
310.10user 5.58system 0:36.41elapsed 866%CPU (0avgtext+0avgdata 675936maxresident)k
0inputs+23672outputs (0major+1565916minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack-cli --stats=errors-only
312.80user 5.57system 0:36.14elapsed 880%CPU (0avgtext+0avgdata 674420maxresident)k
0inputs+23672outputs (0major+1549756minor)pagefaults 0swaps

**Comment by alexander-akait** [comment]
@andersk What is Node.js version, built-in `worker_threads` is faster, but there are some regressions in Node.js in some versions (especial on old 10 and 12)

**Comment by andersk** [comment]
14.17.0 (as specified in my initial report).

I can reproduce the same issue in Node.js 10.24.1 with `NODE_OPTIONS=--experimental-worker`, 12.22.1, 14.17.0, and 16.3.0.

| Node.js version | stock | remove `enableWorkerThreads` |
| --- | --- | --- |
| 10.24.1 | 574900 KiB | 583024 KiB |
| 10.24.1 `--experimental-worker` | 3836652 KiB | 596276 KiB |
| 12.22.1 | 3625192 KiB | 672796 KiB |
| 14.17.0 | 3620284 KiB | 673044 KiB |
| 16.3.0 | 3940128 KiB | 557992 KiB |

If there’s a speed difference, it’s not perceptible.

| Node.js version | stock | remove `enableWorkerThreads` |
| --- | --- | --- |
| 10.24.1 | 36.78s | 36.92s |
| 10.24.1 `--experimental-worker` | 37.49s | 36.97s |
| 12.22.1 | 36.95s | 37.16s |
| 14.17.0 | 36.43s | 36.88s
| 16.3.0 | 36.57s | 37.88s |

**Comment by alexander-akait** [comment]
@andersk Can you check `5.1.1` version?

@SimenB maybe you have understand why this is happening, we use `jest-worker` in `terser-webpack-plugin` and enable `enableWorkerThreads`, but memory consuming is very big

**Comment by alexander-akait** [comment]
We use `jest-worker` v26 previously (https://github.com/webpack-contrib/terser-webpack-plugin/blob/v5.1.2/package.json#L47), then we update `jest-worker` to v27 (https://github.com/webpack-contrib/terser-webpack-plugin/blob/v5.1.3/package.json#L47) and look somewhere memory leak in `jest-worker`...

**Comment by alan-agius4** [comment]
While I don't know if there is a memory regression in jest-worker version 27.

Increase in memory usage in the main process is expected when using worker threads. This is because unlike forked processes work thread are same process. Overall though, the memory consumption should be lower because there is no overhead from the extra forked processes.

**Comment by alexander-akait** [comment]
@alan-agius4 I think so because only jest-worker will be updated in near future, will be great of somebody check it on the big project

**Comment by sergeu90** [comment]
On the 5.1.1 works good on the 5.1.2 i get error. vm have 2gb ram.

**Comment by andersk** [comment]
> @andersk Can you check `5.1.1` version (`terser-webpack-plugin`)?

For the test case I’ve shown, the problem seems the same with terser-webpack-plugin@5.1.1.

<details>
<summary>yarn.lock patch</summary>

```diff
--- a/yarn.lock
+++ b/yarn.lock
@@ -2637,14 +2637,14 @@ isobject@^3.0.1:
   resolved "https://registry.yarnpkg.com/isobject/-/isobject-3.0.1.tgz#4e431e92b11a9731636aa1f9c8d1ccbcfdab78df"
   integrity sha1-TkMekrEalzFjaqH5yNHMvP2reN8=
 
-jest-worker@^27.0.2:
-  version "27.0.2"
-  resolved "https://registry.yarnpkg.com/jest-worker/-/jest-worker-27.0.2.tgz#4ebeb56cef48b3e7514552f80d0d80c0129f0b05"
-  integrity sha512-EoBdilOTTyOgmHXtw/cPc+ZrCA0KJMrkXzkrPGNwLmnvvlN1nj7MPrxpT7m+otSv2e1TLaVffzDnE/LB14zJMg==
+jest-worker@^26.6.2:
+  version "26.6.2"
+  resolved "https://registry.yarnpkg.com/jest-worker/-/jest-worker-26.6.2.tgz#7f72cbc4d643c365e27b9fd775f9d0eaa9c7a8ed"
+  integrity sha512-KWYVV1c4i+jbMpaBC+U++4Va0cp8OisU185o73T1vo99hqi7w8tSJfUXYswwqqrjzwxa6KpRK54WhPvwf5w6PQ==
   dependencies:
     "@types/node" "*"
     merge-stream "^2.0.0"
-    supports-color "^8.0.0"
+    supports-color "^7.0.0"
 
 jquery-caret-plugin@^1.5.2:
   version "1.5.3"
@@ -4053,10 +4053,10 @@ supports-color@^5.3.0:
   dependencies:
     has-flag "^3.0.0"
 
-supports-color@^8.0.0:
-  version "8.1.1"
-  resolved "https://registry.yarnpkg.com/supports-color/-/supports-color-8.1.1.tgz#cd6fc17e28500cff56c1b86c0a7fd4a54a73005c"
-  integrity sha512-MpUEN2OodtUzxvKQl72cUF7RQ5EiHsGvSsVG0ia9c5RbWGL2CI4C7EpPS8UTBIplnlzZiNuV56w+FuNxy3ty2Q==
+supports-color@^7.0.0:
+  version "7.2.0"
+  resolved "https://registry.yarnpkg.com/supports-color/-/supports-color-7.2.0.tgz#1b7dcdcb32b8138801b3e478ba6a51caa89648da"
+  integrity sha512-qpCAvRl9stuOHveKsn7HncJRvv501qIacKzQlO/+Lwxc9+0q2wLyv4Dfvt80/DPn2pqOBsJdDiogXGR9+OvwRw==
   dependencies:
     has-flag "^4.0.0"
 
@@ -4101,18 +4101,18 @@ tapable@^2.1.1, tapable@^2.2.0:
   integrity sha512-FBk4IesMV1rBxX2tfiK8RAmogtWn53puLOQlvO8XuwlgxcYbP4mVPS9Ph4aeamSyyVjOl24aYWAuc8U5kCVwMw==
 
 terser-webpack-plugin@^5.1.1:
-  version "5.1.3"
-  resolved "https://registry.yarnpkg.com/terser-webpack-plugin/-/terser-webpack-plugin-5.1.3.tgz#30033e955ca28b55664f1e4b30a1347e61aa23af"
-  integrity sha512-cxGbMqr6+A2hrIB5ehFIF+F/iST5ZOxvOmy9zih9ySbP1C2oEWQSOUS+2SNBTjzx5xLKO4xnod9eywdfq1Nb9A==
+  version "5.1.1"
+  resolved "https://registry.yarnpkg.com/terser-webpack-plugin/-/terser-webpack-plugin-5.1.1.tgz#7effadee06f7ecfa093dbbd3e9ab23f5f3ed8673"
+  integrity sha512-5XNNXZiR8YO6X6KhSGXfY0QrGrCRlSwAEjIIrlRQR4W8nP69TaJUlh3bkuac6zzgspiGPfKEHcY295MMVExl5Q==
   dependencies:
-    jest-worker "^27.0.2"
+    jest-worker "^26.6.2"
     p-limit "^3.1.0"
     schema-utils "^3.0.0"
     serialize-javascript "^5.0.1"
     source-map "^0.6.1"
-    terser "^5.7.0"
+    terser "^5.5.1"
 
-terser@^5.7.0:
+terser@^5.5.1:
   version "5.7.0"
   resolved "https://registry.yarnpkg.com/terser/-/terser-5.7.0.tgz#a761eeec206bc87b605ab13029876ead938ae693"
   integrity sha512-HP5/9hp2UaZt5fYkuhNBR8YyRcT8juw8+uFbAme53iN9hblvKnLUTKkmwJG6ocWpIKf8UK4DoeWG4ty0J6S6/g==
```

</details>

```console
$ yarn list --pattern=terser-webpack-plugin
yarn list v1.22.10
└─ terser-webpack-plugin@5.1.1
Done in 0.36s.
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
313.69user 3.85system 0:36.30elapsed 874%CPU (0avgtext+0avgdata 3798732maxresident)k
0inputs+23704outputs (0major+1243086minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
313.42user 3.77system 0:36.59elapsed 866%CPU (0avgtext+0avgdata 3479656maxresident)k
0inputs+23704outputs (0major+1221138minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
311.58user 3.83system 0:35.92elapsed 877%CPU (0avgtext+0avgdata 3910884maxresident)k
0inputs+23704outputs (0major+1287653minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
310.42user 3.85system 0:35.98elapsed 873%CPU (0avgtext+0avgdata 3721268maxresident)k
0inputs+23672outputs (0major+1251587minor)pagefaults 0swaps
$ rm -rf dist node_modules/.cache; \time node_modules/.bin/webpack --stats=errors-only
311.79user 3.84system 0:36.29elapsed 869%CPU (0avgtext+0avgdata 3683624maxresident)k
0inputs+23704outputs (0major+1263828minor)pagefaults 0swaps
```

**Comment by alexander-akait** [comment]
@andersk Maybe you can share your application with me (if need I can sign NDA)? I can disable `enableWorkerThreads`, but I feel it is wrong, will be great to investigate why it happens

**Comment by sokra** [comment]
It looks very much like what @alan-agius4 said:

time doesn't measure memory usage of child processes, so using worker_threads will report higher memory usage.

**Comment by andersk** [comment]
@alexander-akait I did share the application and a full reproduction recipe in the original report above. Totally open source, Apache 2.0 license, no NDA necessary. ☺

@sokra Hmm. You seem to be right that `time` doesn’t measure what I thought it was measuring. It does measure over the entire child process tree, but it only reports the largest *single* process within that tree, rather than the total usage of all processes.

I found a tool [cgmemtime](https://github.com/gsauthof/cgmemtime) that does measure the total usage. Based on that better measurement, the discrepancy between webpack 4 and webpack 5 still exists in the actual application, but it’s less dramatic, and doesn’t seem to be visible in this particular benchmark that I extracted.

I will close this report for now, do more experiments with cgmemtime, and open another report if I find something actionable. Feel free to hop into zulip/zulip#16649 if you’d like to follow along.

Thanks all!
